sam.c

```diff
@@ -2383,9 +2383,175 @@ int sam_hdr_change_HD(sam_hdr_t *h, const char *key, const char *val)
  *** SAM record I/O ***
  **********************/
 
-static int sam_parse_B_vals(char type, uint32_t n, char *in, char **end,
-                            char *r, bam1_t *b)
-{
+// The speed of this code can vary considerably depending on minor code
+// changes elsewhere as some of the tight loops are particularly prone to
+// speed changes when the instruction blocks are split over a 32-byte
+// boundary.  To protect against this, we explicitly specify an alignment
+// for this function.  If this is insufficient, we may also wish to
+// consider alignment of blocks within this function via
+// __attribute__((optimize("align-loops=5"))) (gcc) or clang equivalents.
+// However it's not very portable.
+// Instead we break into separate functions so we can explicitly specify
+// use __attribute__((aligned(32))) instead and force consistent loop
+// alignment.
+static inline int64_t grow_B_array(bam1_t *b, uint32_t *n, size_t size) {
+    // Avoid overflow on 32-bit platforms, but it breaks BAM anyway
+    if (size > INT32_MAX*0.666) {
```

Comment:
I just noticed that this test does nothing, as `size` is never greater than 4.  It probably need to be something like this to trap the wrap-around case:

```suggestion
    if (size > 2 && (*n) > SIZE_MAX/2) {
```

Anything else will be dealt with by `possibly_expand_bam_data()`.

Reply:
Oops, indeed.   Confusing myself with the meaning of variables.

Reply:
I changed this to `if (*n > INT32_MAX*0.666)` as the BAM limit in the B array is the 32-bit number of elements and this is independent on the size of elements.  I just used the wrong field.  `SIZE_MAX` is also the wrong type as I'm not worried about overflowing size_t, but the BAM 32-bit limitation.

The *0.666 is because we basically multiply n by 1.5 each time we grow it. Ie n += (n>>1).

Reply:
I was more worried about the line:
```
size_t bytes = (size_t)size * (size_t)(*n>>1);
```
as if `size == 4`, `bytes` is set to `2 * (*n)`.  I'll grant that the number of elements in the B array is also a problem though.  Luckily you compare to `INT32_MAX` so it works for both.